java/query/src/main/java/sleeper/query/model/QueryJson.java

```diff
@@ -63,60 +65,81 @@ static QueryJson from(Query query, QuerySerDe.SchemaLoader schemaLoader) {
+                .subQueryId(leafQuery.getSubQueryId())
+                .subQueryRegions(leafQuery.getRegions().stream()
+                        .map(regionSerDe::toJsonTree)
+                        .collect(Collectors.toUnmodifiableList()))
+                .leafPartitionId(leafQuery.getLeafPartitionId())
+                .partitionRegion(regionSerDe.toJsonTree(leafQuery.getPartitionRegion()))
+                .files(leafQuery.getFiles())
+                .build();
+    }
+
+    QueryOrLeafQuery toQueryOrLeafQuery(QuerySerDe.SchemaLoader schemaLoader) {
+        validate();
+        RegionSerDe regionSerDe = regionSerDe(schemaLoader);
         switch (type) {
             case "Query":
-                return new Query.Builder(tableName, queryId, readRegions(regionSerDe))
-                        .setRequestedValueFields(requestedValueFields)
-                        .setQueryTimeIteratorClassName(queryTimeIteratorClassName)
-                        .setQueryTimeIteratorConfig(queryTimeIteratorConfig)
-                        .setResultsPublisherConfig(resultsPublisherConfig)
-                        .setStatusReportDestinations(readStatusReportDestinations())
-                        .build();
+                return new QueryOrLeafQuery(toParentQuery(regionSerDe));
             case "LeafPartitionQuery":
-                Region partitionRegion = regionSerDe.fromJsonTree(this.partitionRegion);
-                return new LeafPartitionQuery.Builder(
-                        tableName, queryId, subQueryId, readRegions(regionSerDe), leafPartitionId, partitionRegion, files)
-                        .setRequestedValueFields(requestedValueFields)
-                        .setQueryTimeIteratorClassName(queryTimeIteratorClassName)
-                        .setQueryTimeIteratorConfig(queryTimeIteratorConfig)
-                        .setResultsPublisherConfig(resultsPublisherConfig)
-                        .setStatusReportDestinations(readStatusReportDestinations())
-                        .build();
+                return new QueryOrLeafQuery(toLeafQuery(regionSerDe));
             default:
                 throw new QueryValidationException(queryId, statusReportDestinations, "Unknown query type \"" + type + "\"");
         }
     }
 
-    private static QueryJson from(LeafPartitionQuery query, RegionSerDe regionSerDe) {
-        return builder(query, regionSerDe)
-                .type("LeafPartitionQuery")
-                .subQueryId(query.getSubQueryId())
-                .leafPartitionId(query.getLeafPartitionId())
-                .partitionRegion(regionSerDe.toJsonTree(query.getPartitionRegion()))
-                .files(query.getFiles())
+    Query toParentQuery(QuerySerDe.SchemaLoader schemaLoader) {
+        validate();
+        return toParentQuery(regionSerDe(schemaLoader));
+    }
+
+    private Query toParentQuery(RegionSerDe regionSerDe) {
+        return new Query.Builder(tableName, queryId, readRegions(regions, regionSerDe))
```

Comment:
Is there a reason we're leaving this odd Query.Builder constructor in, as opposed to converting it to a proper builder like you've done with `LeafPartitionQuery`?

Reply:
Just to limit the size of the pull request. I've started making those changes in https://github.com/gchq/sleeper/pull/1490.